extensions/2.0/Vendor/EXT_property_animation/README.md

```diff
@@ -0,0 +1,111 @@
+# EXT\_property\_animation
+
+## Contributors
+
+* Bryce Hutchings, Microsoft [@brycehutchings](https://twitter.com/brycehutchings)
+* Gary Hsu, Microsoft [@bghgary](https://twitter.com/bghgary)
+* Jamie Marconi, Microsoft [@najadojo](https://twitter.com/najadojo)
+* Lewis Weaver, Microsoft
+
+## Status
+
+Draft
+
+## Dependencies
+
+Written against the glTF 2.0 spec.
+
+## Overview
+
+This extension adds a specification for animation data targeting arbitrary
+properties, such as material colors, texture transform matrices and
+extension properties.
+
+## Extending Animations
+
+Property animations can be added to an animation by adding the
+`EXT_property_animation` extension to any glTF animation. For example, the
+following defines an animation with two channels that modify a material's
+baseColorFactor and roughnessFactor.
+```json
+"animations" : [
+    {
+        "channels" : [], 
```

Comment:
Empty lists are forbidden by the current schema. Why not just re-use the existing `target` property (since the schema allows arbitrary strings there)?

Reply:
I updated the sample such that it doesn't fall into this problem. We can't use the `path` as it exists because its defined as an enum value with only the four existing supported values. 

There may be assets that wish only to animate some property values but not TRS or weights. Suggestions welcome for schema change or a best practice note for an identity animation that would go in core spec `channels`.

Reply:
The schema allows any string in addition to pre-defined four values. 
https://github.com/KhronosGroup/glTF/blob/09a404da7459f23b0a79c98c1afdb26cdad57f32/specification/2.0/schema/animation.channel.target.schema.json#L27-L29

Reply:
It does but the validator and more than one engine makes this a strict enum value. Plus this schema requires a node reference which isn't required with the JSON Pointer.

Reply:
Node reference is not required
https://github.com/KhronosGroup/glTF/blob/09a404da7459f23b0a79c98c1afdb26cdad57f32/specification/2.0/schema/animation.channel.target.schema.json#L35

The latest validator revision (wip) no longer treats unknown enum values as errors. 

The documentation about extensions explicitly says that they can extend allowed enums.

Reply:
Oh! That would certainly solve the problem. So the suggestion is that this be modified this an empty target extension:
```json
"channels" : [
    {
        "sampler" : 0, 
        "target" : {
            "path" : "/materials/1/pbrMetallicRoughness/roughnessFactor",
            "extensions" : {
                "EXT_property_animation" : {}
            }
        }
    }
], 
```
? It could get a bit wordy in the JSON but I do like it better.

Reply:
Even the extension object isn't needed. Top-level `extensionsUsed` declaration should be enough schema-wise. 

E.g., see how `MSFT_texture_dds` extends `image.mimeType` with `image/vnd-ms.dds`.

Reply:
I need to see an example to be sure, but I think doing it without an extension isn't forward-compatible. It works for MSFT_texture_dds because implementations without the extension will not try to load the textures. The same is not true for this extension.

Reply:
"forward-compatible" isn't quite the right phrase for this since it's not a new version. What I mean is that loaders that don't support this extension will not be able to load an asset with this extension (if we add new types to the channels without using another list).

Reply:
What do you mean by "new channels"? JSON objects with this extension remain within existing data types.

Reply:
Sorry, I meant new path enumeration type. An existing implementation will fail to load if a new path is now part of the channels list.

Reply:
It also seems node is not optional for existing engines.

Reply:
AFAIR, this specific use case (extensibility of animation targets) has been the reason why `node` is optional in the schema.

Reply:
Leaving this extension aside, implementations should be able to handle unknown enums gracefully since that's why all enums have been extended with `"type": "string"`-like alternatives.

If we cannot use existing `animation.channel.target` objects for some app-internal reasons, a new top-level array seems to be the only remaining option for this extension to not break the schema.

Reply:
> Leaving this extension aside, implementations should be able to handle unknown enums gracefully since that's why all enums have been extended with `"type": "string"`-like alternatives.

I'm not sure. For animation paths, it's easy to gracefully ignore them, but I don't think it will be easy all the time. For example, if we introduce a new accessor type, does that mean the whole mesh should not load if one of accessors of a primitives of the mesh uses it?

We talked about this a lot for versioning [here](https://github.com/KhronosGroup/glTF/pull/899#issuecomment-292678720). This is very similar to versioning in a lot of ways.

Reply:
Yeah I remember, but it seems like we're experiencing practical implications of it only now.

We may choose to limit enums extensibility strictly to `minVersion`/`extensionsRequired` case (but `MSFT_texture_dds` doesn't follow that with `mimeType`). 

It'd useful to have a consistent way of dealing with enums rather than returning to this on a case-by case basis.

Reply:
> It'd be useful to have a consistent way of dealing with enums rather than returning to this on a case-by case basis.

I think there are two cases:
1. An object is not referenced unless an extension is processed. This object can have extended enumeration types.
2. An object is referenced by the core spec. This object must use a separate property not in the core spec to extend functionality.

Reply:
I see how it may work although this distinction complicates promotion a bit (so schemas must be remade for minor version updates).

Looking again at the animation example above, it seems that `target` isn't referenced by any other property but inlined into `channel`, so it's neither case, right?

Reply:
> schemas must be remade for minor version updates

I'm not sure what you mean. What schema changes are you talking about?

> Looking again at the animation example above, it seems that `target` isn't referenced by any other property but inlined into `channel`, so it's neither case, right?

`target` is being referenced through animations[n].channels[m].target. Basically, any enum that will be parsed and interpreted by an implementation loading the core spec falls into case 2.

Reply:
> What schema changes are you talking about?

Updated properties reside in the extension object for now. If/when the same functionality goes into `2.x` core version, the schema will be different. That would complicate implementations willing to support all combinations of core versions / extensions.

> `target` is being referenced through animations[n].channels[m].target.

I'd say that `target` is contained within `channel` object, not referenced by it. So an engine may transitively treat the whole `channel` object as "unknown" as well as `animation`. Given that  there're no external (indexed) references to any of animation-related objects, I don't see much difference with top-level `image` with updated `mimeType`.

Reply:
> I don't see much difference with top-level `image` with updated `mimeType`.

I suppose it depends on implementation. If an implementation is a top-down loader, then I see a difference. Images referenced only by an extension will never be loaded if the extension is never read and thus the types in the image can be extended without fear of existing implementations falling over.

Reply:
> I suppose it depends on implementation.

Exactly. Given that the spec allows an asset to have only, e.g., `images` collection, we can't assume that an arbitrary implementation won't try to load them in advance unless we want to incorporate "top-down" approach into spec's compatibility strategy (which may be a good idea after all).

The existing schema prevents us from extending `animation.channel.target` with an extension object because `animation.channel.target.path` is required. With versioning concerns in mind, this means that the safest way of designing the extension schema is to use a new top-level collection.